text/0000-macros-1.1.md

```diff
@@ -0,0 +1,552 @@
+copy/pasted. All span information for the `TokenStream` structures returned by
+`from_source` will point to the original `#[derive]` annotation. This means
+that error messages related to struct definitions will get *worse* if they have
+a custom derive attribute placed on them, because the entire struct's span will
+get folded into the `#[derive]` annotation. Eventually, though, more span
+information will be stable on the `TokenStream` type, so this is just a
+temporary limitation.
+
+The `rustc_macro_derive` attribute requires the signature (similar to [macros
+2.0][mac20sig]):
+
+[mac20sig]: http://ncameron.org/blog/libmacro/#tokenisingandquasiquoting
+
+```rust
+fn(&mut Context, TokenStream) -> TokenStream
+```
+
+If a macro cannot process the input token stream, it is expected to panic for
+now, although eventually it will call methods on `Context` to provide more
+structured errors. The compiler will wrap up the panic message and display it
+to the user appropriately. Eventually, however, `librustc_macro` will provide
+more interesting methods of signaling errors to users.
+
+Customization of user-defined `#[derive]` modes can still be done through custom
+attributes, although it will be required for `rustc_macro_derive`
+implementations to remove these attributes when handing them back to the
+compiler. The compiler will still gate unknown attributes by default.
+
+### `rustc-macro` crates
+
+Like the executable, staticlib, and cdylib crate types, the `rustc-macro` crate
+type is intended to be a final product.  What it *actually* produces is not
+specified, but if a `-L` path is provided to it then the compiler will recognize
+the output artifacts as a macro and it can be loaded for a program.
+
+Initially if a crate is compiled with the `rustc-macro` crate type (and possibly
+others) it will forbid exporting any items in the crate other than those
+functions tagged `#[rustc_macro_derive]` and those functions must also be placed
+at the crate root. Finally, the compiler will automatically set the
+`cfg(rustc_macro)` annotation whenever any crate type of a compilation is the
+`rustc-macro` crate type.
+
+While these properties may seem a bit odd, they're intended to allow a number of
+forwards-compatible extensions to be implemented in macros 2.0:
+
+* Macros eventually want to be imported from crates (e.g. `use foo::bar!`) and
+  limiting where `#[derive]` can be defined reduces the surface area for
+  possible conflict.
+* Macro crates eventually want to be compiled to be available both at runtime
+  and at compile time. That is, an `extern crate foo` annotation may load
+  *both* a `rustc-macro` crate and a crate to link against, if they are
+  available. Limiting the public exports for now to only custom-derive
+  annotations should allow for maximal flexibility here.
+
+### Using a procedural macro
+
+Using a procedural macro will be very similar to today's `extern crate` system,
+such as:
+
+```rust
+extern crate double;
```

Comment:
The problem is that if you do *anything* by default now, you'll need to opt out to be able to use the normal module system.